File: src/core/observable/SimpleEmitter.ts
import { Disposable } from "./Disposable";
import { Emitter } from "./Emitter";
import { ChangeEvent } from "./Observable";
import { LogManager } from "../Logger";

const logger = LogManager.get("core/observable/SimpleEmitter");

export class SimpleEmitter<T> implements Emitter<T> {
    protected readonly observers: ((event: ChangeEvent<T>) => void)[] = [];

    emit(event: ChangeEvent<T>): void {
        for (const observer of this.observers) {
            try {
                observer(event);
            } catch (e) {
                logger.error("Observer threw error.", e);
            }
        }
    }

    observe(observer: (event: ChangeEvent<T>) => void): Disposable {
        if (!this.observers.includes(observer)) {
            this.observers.push(observer);
        }

        return {
            dispose: () => {
                const index = this.observers.indexOf(observer);

                if (index !== -1) {
                    this.observers.splice(index, 1);
                }
            },
        };
    }
}

File: observable/src/commonMain/kotlin/world/phantasmal/observable/SimpleEmitter.kt
package world.phantasmal.observable

import world.phantasmal.core.disposable.Disposable

// TODO: Should multiple events be emitted somehow during a change set? At the moment no application
//       code seems to care.
class SimpleEmitter<T> : AbstractDependency<T>(), Emitter<T> {
    override var changeEvent: ChangeEvent<T>? = null
        private set

    override fun emit(event: ChangeEvent<T>) {
        applyChange {
            this.changeEvent = event
        }
    }

    override fun observeChange(observer: ChangeObserver<T>): Disposable =
        CallbackChangeObserver(this, observer)
}
